src/command/archive/push/file.c

```diff
@@ -163,11 +163,39 @@ archivePushFile(
             const String *const walSegmentChecksum = strNewEncode(
                 encodingHex, pckReadBinP(ioFilterGroupResultP(ioReadFilterGroup(read), CRYPTO_HASH_FILTER_TYPE)));
 
+            // Is the check command currently running?
+            const bool checkLockHeld = lockRead(
+                cfgOptionStr(cfgOptLockPath), cfgOptionStr(cfgOptStanza), lockTypeCheck).status == lockReadStatusValid;
+
+            // Is the backup command running for a specific repo?
+            const Variant *runningBackupRepoIdx = lockRead(
+                cfgOptionStr(cfgOptLockPath), cfgOptionStr(cfgOptStanza), lockTypeBackup).data.repoIdx;
+
             // Check each repo for the WAL segment
             for (unsigned int repoListIdx = 0; repoListIdx < lstSize(repoList); repoListIdx++)
             {
                 const ArchivePushFileRepoData *const repoData = lstGet(repoList, repoListIdx);
 
+                // Skip archiving if archive retention is set to 0 and the backup command isn't currently running for this specific
+                // repository. Archiving should always happen when the check command is running for all the repositories.
+                bool lockHeld = false;
+                if (checkLockHeld || (runningBackupRepoIdx != NULL && varUInt(runningBackupRepoIdx) == repoData->repoIdx))
+                    lockHeld = true;
+
+                if (!lockHeld
+                    && (cfgOptionIdxTest(cfgOptRepoRetentionArchive, repoData->repoIdx)
+                        && cfgOptionIdxUInt(cfgOptRepoRetentionArchive, repoData->repoIdx) == 0))
+                {
+                    // Add warning to the result that will be returned to the main process
+                    strLstAddFmt(
+                        result.warnList,
+                        "WAL file '%s' skipped for %s, because option '%s' is set to 0",
+                        strZ(archiveFile), cfgOptionGroupName(cfgOptGrpRepo, repoData->repoIdx),
+                        cfgOptionIdxName(cfgOptRepoRetentionArchive, repoData->repoIdx));
+                    destinationCopy[repoListIdx] = false;
+                    continue;
+                }
```

Comment:
We currently not report any missing wal file in the info command (nor in the verify command the last time I tried). So perhaps not warning but perhaps detail (just like we have for the copied files in the backup command)?

Reply:
Changed to detail in [3b806c6](https://github.com/pgbackrest/pgbackrest/pull/1926/commits/3b806c6233f14888c44cde32de34c7455f24562e). 
This seems important to note that the .backup files aren't pushed either :thinking:

Reply:
.backup files are pretty useless so I'm OK with that. One problem: this detail messages is not going to make it into the main log. It will only show up when logging with --log-subprocess. This is why the warnings are passed back to the main process to be output there.

Reply:
Hm, that's weird because it's not what I saw during the tests (explicitly setting `--no-log-subprocess`) :thinking: 

```
P00   INFO: archive-push command begin 2.44dev: [pg_wal/000000010000000000000016] --exec-id=37899-96f2d7d8 --log-level-console=detail --log-level-file=detail --no-log-subprocess --pg1-path=/var/lib/pgsql/15/main/data --process-max=2 --repo1-cipher-pass=<redacted> --repo2-cipher-pass=<redacted> --repo1-cipher-type=aes-256-cbc --repo2-cipher-type=aes-256-cbc --repo1-host=172.18.0.2 --repo2-host=172.18.0.2 --repo1-host-user=pgbackrest --repo2-host-user=pgbackrest --repo1-path=/repo1 --repo2-path=/repo2 --repo1-retention-archive=0 --stanza=ro8pg

P00 DETAIL: '000000010000000000000016' skipped for repo1, because option 'repo1-retention-archive' is set to 0
```

Reply:
It has to do with the way local processes are forked off in unit tests. It works differently when running pgbackrest for real, i.e. you won't get the logs for the subprocess in the main log.